### ISSUE TITLE
Allow functions with optional arguments as generics

### ISSUE THREAD

**Issue by catwell** [teal-language/tl#281]
For instance:

```
local function map<A,B>(t: {A}, f: function(A):(B)): {B}
    local r = {}
    for i, v in ipairs(t) do
        r[i] = f(v)
    end
    return r
end

local function f(a: number, b: number) : number
    return a * (b or 2)
end

print(map({1,2,3}, f)[2])
```

To be coherent with [this](https://github.com/teal-language/tl/issues/71#issuecomment-707804167) it should work, but instead it prints:

```
x.tl:13:20: argument 2: incompatible number of arguments: got 2 (number, number), expected 1 (A)
```

**Comment by hishamhm** [comment]
@catwell This is a super interesting topic! We had [an interesting discussion about this issue](https://www.twitch.tv/videos/843366632) on my Twitch stream last Saturday.

The summary of it is that currently, Teal is strict about arity when matching type declarations (because type declarations are a new Teal-only concept, so we can afford to be strict), but it is lax about arity when doing function calls (because the behavior of function calls are inherited from Lua).

As seen in #71, we would like to find a way to be strict about arity in function calls as well (but that then also touches on the trickier topic of nullability).

I'm open for changing my mind, but as of today, I don't think the example above in this issue should work. I think we should require a cast on `f`. If we add a way to specify that argument `b` in function `f` is optional, then I think this should work.

I'll keep this open and label it as "enhancement"!

**Comment by catwell** [comment]
I agree, a way to specify optional arguments as described in the other issue would be even better (and if function calls are made stricter it would help with error detection before runtime).
